### PR TITLE
[WIP] Random lpca

### PR DESCRIPTION
**Pull request by 15thai**
New Implementation - Random matrix local pca. 
References    
    ----------
    .. [Veraart16] Veraart J, Fieremans E, Novikov DS (2016)
                  Diffusion MRI noise mapping using random matrix theory.
                  Magnetic resonance in Medicine 76(5), p1582-1593.
                  https://doi.org/10.1002/mrm.26059
    """

### CODE REVIEW

File: dipy/denoise/randommatrix_localpca_parallel.pyx [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+              
+                else:
+                    for ii in range(rr):
+                        temp1[k,ii]= C[k,ii,nn/2]
+
+                    fast_matvec('n',diag_eigmat[k,:,:], temp1[k,:] ,  temp0[k,:])
+                    fast_matvec('n',C[k,:,:], temp0[k,:],   temp1[k,:])
+                    fast_matvec('n',X[k,:,:], temp1[k,:],   temp2[k,:])
+                
+                denoised_arr_view[i,j,k,:] = temp2[k,:]
```
Comment (RafaelNH) [inline dipy/denoise/randommatrix_localpca_parallel.pyx]: You are only using the center voxel of each patch for denoising. Please consider using all voxels like using an overcomplete average approach (Manjon et al., 2013)[https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0073021]. In this way, we avoid removing information in cases that tissues are near to the limits of the image or in cases that patch size is large. Keep in mind that some researcher might use PCA in all voxels of an image simultaneously

File: dipy/denoise/randommatrix_localpca_parallel.pyx [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        dgemm(&transa, &transb, &m , &n, &k, &alpha, a0, &m, a0,
+               &n, &beta, c0, &m)  
+       
+    return 0
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+def randommatrix_localpca_parallel(arr, patch_extent=0, out_dtype=None,num_threads=None):
```
Comment (RafaelNH) [inline dipy/denoise/randommatrix_localpca_parallel.pyx]: Can you remake this function to something shorter? Suggestion: pca_denoise.

**Comment by RafaelNH** [comment]
Hi @15thai! Very nice start!
See my two comments in the code. In addition to those, please reformat the code using PEP8 style. Do not forget to add tests and an example for the code usage. Keep the good work!

File: dipy/denoise/randommatrix_localpca_parallel.pyx (hunk 2 above)
Comment (ShreyasFadnavis) [reply dipy/denoise/randommatrix_localpca_parallel.pyx]: Hi Anh! I agree with @RafaelNH on this one, I would suggest randommatrix_lpca.pyx

**Comment by 15thai** [comment]
Hey. Sorry for such a delayed work. I am deeply sorry. We are still working on with the C++. I have a curious question if I am not implementing the overcomplete. What if we are using padding zero, and Eigen Decomposition for the border of the image. Will it be possible? Let me know your thought. Thank you again.

**Comment by RafaelNH** [comment]
Hi @15thai! Can you give more details on what you want to do? I was suggesting to do the overcomplete averaging so that denoising of the border image is done in a continuous way relative to the centre voxels. This strategy is already used in the current version of dipy's PCA denoising - you can see how that is done in lines 156-158 in localpca.py.

**Comment by 15thai** [comment]
@eurotomania Can you see this ? Here is the PR in the DIPY group. 
I want to include Okan in, so he can understand and see the process. Thank you.

**Comment by ShreyasFadnavis** [comment]
@15thai something went wrong with Git! Can you try to `revert` to commit `ce41956`? Or `rebase` maybe?
@skoudoro can you check this?

**Comment by skoudoro** [comment]
I will try asap

**Comment by 15thai** [comment]
I misunderstood the merging my local to my branch. What should I do?

**Pull request closed without merging**